lnwallet/chancloser/chancloser.go

```diff
@@ -470,6 +476,373 @@ func validateShutdownScript(disconnect func() error, upfrontScript,
+		if err := validateShutdownScript(
+			c.cfg.Disconnect,
+			c.cfg.Channel.RemoteUpfrontShutdownScript(),
+			msg.Address, c.cfg.ChainParams,
+		); err != nil {
+			return noShutdown, err
+		}
+
+		// Once we have checked that the other party has not violated
+		// option upfront shutdown we set their preference for delivery
+		// address. We'll use this when we craft the closure
+		// transaction.
+		c.remoteDeliveryScript = msg.Address
+
+		// We'll generate a shutdown message of our own to send across
+		// the wire.
+		localShutdown, err := c.initChanShutdown()
+		if err != nil {
+			return noShutdown, err
+		}
+
+		// If this is a taproot channel, then we'll want to stash the
+		// remote nonces so we can properly create a new musig
+		// session for signing.
+		if c.cfg.Channel.ChanType().IsTaproot() {
+			if msg.ShutdownNonce == nil {
+				return noShutdown,
+					fmt.Errorf(
+						"shutdown nonce not populated",
+					)
+			}
+
+			c.cfg.MusigSession.InitRemoteNonce(&musig2.Nonces{
+				PubNonce: *msg.ShutdownNonce,
+			})
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): responding to shutdown",
+			c.chanPoint)
+
+		// After the other party receives this message, we'll actually
+		// start the final stage of the closure process: fee
+		// negotiation. So we'll update our internal state to reflect
+		// this, so we can handle the next message sent.
+		c.state = closeAwaitingFlush
+
+		return fn.Some(*localShutdown), err
+
+	case closeShutdownInitiated:
+		// If the remote node opened the channel with option upfront
+		// shutdown script, check that the script they provided matches.
+		if err := validateShutdownScript(
+			c.cfg.Disconnect,
+			c.cfg.Channel.RemoteUpfrontShutdownScript(),
+			msg.Address,
+			c.cfg.ChainParams,
+		); err != nil {
+			return noShutdown, err
+		}
+
+		// Now that we know this is a valid shutdown message and
+		// address, we'll record their preferred delivery closing
+		// script.
+		c.remoteDeliveryScript = msg.Address
+
+		// At this point, we can now start the fee negotiation state, by
+		// constructing and sending our initial signature for what we
+		// think the closing transaction should look like.
+		c.state = closeAwaitingFlush
+
+		// If this is a taproot channel, then we'll want to stash the
+		// local+remote nonces so we can properly create a new musig
+		// session for signing.
+		if c.cfg.Channel.ChanType().IsTaproot() {
+			if msg.ShutdownNonce == nil {
+				return noShutdown,
+					fmt.Errorf("shutdown " +
+						"nonce not populated")
+			}
+
+			c.cfg.MusigSession.InitRemoteNonce(&musig2.Nonces{
+				PubNonce: *msg.ShutdownNonce,
+			})
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): shutdown response "+
+			"received, entering fee negotiation", c.chanPoint)
+
+		return noShutdown, nil
+
+	default:
+		// Otherwise we are not in a state where we can accept this
+		// message.
+		return noShutdown, ErrInvalidState
+	}
+}
+
+func (c *ChanCloser) BeginNegotiation() (
```

Comment:
Missing `godoc` comments for these. Generally I think we can have the best of worlds if: all fo these new methods are made private, they're called from the existing `ProcessCloseMsg` dispatch func.

Reply:
Commit wise, I think if each commit takes a chunk from `ProcessCloseMsg` and refactors it into one of these new messages (vs adding the new func that may or may not contain new logic), then it'll be easier to review.

Reply:
> all fo these new methods are made private, they're called from the existing ProcessCloseMsg dispatch func.

Which CloseMsg corresponds to `BeginNegotiation`?